### PR TITLE
Create Using_Linux_in_a_Safe_System.md

### PR DESCRIPTION
**Pull request by igor-stoppa**
Considerations guiding the design of a safe system including Linux.

### CODE REVIEW

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+  satisfied through some component other than Linux (more on this later).
+
+  There are several additional reasons for favouring an external component, besides
+  what has been mentioned already:
+  *  much greater probability for the component to be simpler (and thus easier to qualify)
+  *  opportunity to have a better control over the component (e.g. use a software
+     architecture that is better suited for safety, formalised processes, etc.)
+  *  decouple release cycles of the Linux kernel from the safety component
+  *  other non-technical reasons (e.g. different licensing)
+
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:113]: Here I would add a sentence saying "indeed the drawback is that the more the components that are present in the system the higher the cost to be paid is"

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+## **Examples of Safety-oriented System Design Concepts**
+
+**Premise:** applying the initial consideration from the previous sections, it is
+necessary to define what is required at system level in terms of both availability
+and integrity.
+
+The simpler - and most common - case is to require detection of interference
+affecting components that are recipient of safety requirements.
+
+Interference affecting a component with safety requirements must be both detected
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:123]: in principle interference is either avoided or detected....so maybe for completeness the avoidance case should be added

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+  Monitoring is a simple approach to detection of interference that affects
+  component with safety requirement.
+  
+  It can save from extensive hardening.
+
+  However two conditions must be met:
+  1. It must be possible to express a simple check, in a way that can repeatedly,
+     reliably and unequivocally produce a YES/NO verdict about the health parameter
+     of interest
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:142]: below jumping from 1. to 3. ....

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+  3. The observation whether the condition is met (or not) can be performed from a
+     context having same or better safety qualification.
+     And the context is able to take whatever action might be dictated by the
+     specific policy in place.
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:146]: IMO there are some safety principles that are not clear above or missing.
1) The monitor is to be independent from the component being monitored (i.e. the monitored failure modes shall not propagate to the monitor)
2) the monitor shall be qualified up to a safety integrity level equal or better than the one allocated to the component being monitored
3) A safety analysis shall verify the effectiveness of the monitor against the failure modes being monitored

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 1 above)
Comment (igor-stoppa) [reply Contributions/Using_Linux_in_a_Safe_System.md:113]: No, that argumentation is incomplete and misleading, one must consider the total cost of ownership: adding one single component and enabling safety claims on that one can be less costly than not adding components and having to make the existing ones safe.

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 3 above)
Comment (igor-stoppa) [reply Contributions/Using_Linux_in_a_Safe_System.md:142]: in reality the rendering cares only about having a number, but ok

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+* **Application Monitoring**
+
+  Monitoring is a simple approach to detection of interference that affects
+  component with safety requirement.
+  
+  It can save from extensive hardening.
+
+  However two conditions must be met:
+  1. It must be possible to express a simple check, in a way that can repeatedly,
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:140]: I would use systematically instead of repeatedly. Simply because by using repeatedly it seems that we are looking at a periodic checker; whereas there could be mechanisms based on interrupts/exceptions....

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+  Having to simply report failures (within a maximum allotted time - and how much?)
+  usually is significantly simpler than having to guarantee that a certain
+  functionality will not be interrupted.
+
+  Clarifying the scope of the design goes a long way in defining criteria for
+  identifying acceptable solutions.
+
+* **Linux is intrinsically not safe**
+
```
[the remark this answers is not in the document]
Comment (igor-stoppa) [reply Contributions/Using_Linux_in_a_Safe_System.md:86]: Individual components *ARE* subject to safety evaluations; in fact one can purchase components, both harware and software, that come with AoUs and safety qualification ceritificates. 

Linux doesn't do any of that, so I think the statement is correct.

I did not say that Linux  cannot be used in a safe system, or that the presence of Linux in a system makes it impossible to provide safety to the system.

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+3. ***CheckList for Safety Claims on a Generic Linux System***
+4. [CC BY-SA 4.0 Deed | Attribution-ShareAlike 4.0 International | Creative Commons](https://creativecommons.org/licenses/by-sa/4.0/) - <https://creativecommons.org/licenses/by-sa/4.0/> License
+
+
+## **Purpose of the document**
+The considerations provided here are meant to help with designing a safe
+system comprising Linux.
+
+There are many reasons why one might want to use Linux in a system, but its
+safety is certainly not one of them - nor it should not be.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:37]:
```suggestion
safety is certainly not one of them - nor should it be.
```

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 2 above)
Comment (igor-stoppa) [reply Contributions/Using_Linux_in_a_Safe_System.md:123]: I am referring to the case I described above.

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 5 above)
Comment (igor-stoppa) [reply Contributions/Using_Linux_in_a_Safe_System.md:140]: ok

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+  responsible of performing safety relevant operations would ping a watchdog,
+  periodically.
+
+  As long as the application pings the watchdog within the set deadline, the watchdog
+  does not trigger any exceptional operation.
+
+  The example is crude, but it illustrates how - for example - all the internal linux
+  components involved with booting, loading and executing the application are not
+  subject to any safety requirement, because only the final product of their operation
+  is monitored.
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:162]: In the period above you are saying that booting, loading, executing the app are not subject to any safety requirements but you have not described why. So it is too early for a conclusion like this here. Instead I would say:

The example is simple, but it suggests how - for example - all the HW/SW stack involved with booting, loading and executing the application could be
  monitored against time interference deriving from any Linux failure mode or failure modes of the HW supporting the Linux execution.

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+
+  In this case one must accept, though, that safety claims are possible only for
+  what is being monitored, either directly or indirectly.
+
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:166]: Above I would say "safety claims are possible only for freedom from the specific interference deriving from the failure modes being monitored (i.e. temporal interference in this case)"

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+  If this is not acceptable, then additional requirements must be formulated,
+  and additional monitoring (or other means of mitigation) must be deployed.
+
+  With regard to the example, having a monitor that detects excessive skew in
+  the period used by the application to ping, will not ensure that other parts of
+  the application are free from interference and that they will function correctly.
+
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:173]: Above you could more clearly say 

With regard to the example, having a monitor that detects excessive skew in
  the period used by the application to ping, will not ensure that other interference
  modes (e.g. memory corruption) can be prevented or detected

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+  Never the less, carefully choosing how/where to implement the monitor and how to
+  prime it can avoid expensive, tedious, error-prone involvement of large chunks
+  of the kernel architecture in supporting safety claims.
+
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:177]: The sentence above is a bit confusing IMO. Till now you have focused on interfernce, whereas now you seem to move also to other safety claims.
So for example the WTD mechanism can detect:
- any temporal failure mode deriving from interference from components outside those functionally needed to run the safety app
- any temporal failure mode deriving from failure modes of the components functionally supporting the safety app

In principle the failure modes above are different as the map to different safety claims....so what are we focusing on here ?

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+**Premise:** some of the following considerations assume a fairly high level
+of freedom in designing the overall system.
+While one must acknowledge that in practice this freedom will not be always
+available, due to design constraints, it is equally true that attempting to
+cope with these constraints can lead to an even higher total cost of ownership.
+
+* **Clearly define requirements toward system integrity and availability, upfront**
+
+  Having to simply report failures (within a maximum allotted time - and how much?)
+  usually is significantly simpler than having to guarantee that a certain
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:79]: As per my experience a system don't have to guarantee that a certain safety related functionality will not be interrupted. 
If it is not possible to ship a safety related functionality system should move to a different state applying recovery actions.

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+     This is fairly boilerplate information, but it is worth mentioning that
+     safety requirements apply not only to software but to hardware as well.
+     The Watchdog IP must be safety qualified, to be relied on.
+
+     However, for the watchdog to support safety claims, the following must
+     be proven:
+     *  the initialisation of the watchdog can be considered sufficiently safe
+        (same or better safety level than the required one)
+     *  it must be impossible for a single failure in the system to make it
+        ineffective; this case covers:
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:217]: impossible???

I'd rather say "the probably of a failure in the watchdog itself or of a failure propagating from the system to the watchdog shall be coherent with the safety integrity level of the safety claim supported by the watchdog"

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+  While Linux has many remarkable properties, safety is not one of them, which means
+  that it won't be possible to just allocate safety requirements to Linux, and expect
+  them to be automatically satisfied. Quite the opposite, actually.
+
+  Having to cope with safety requirements assigned to Linux will involve a price to
+  be paid, in one form of another.
+
+  The problem with said cost is that it might *also* affect negatively the goals
+  that led to the selection of Linux, in the first place. For example, if Linux was
+  selected because of its a rapid turnaround in fixing security bugs, selecting
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:96]: of its a rapid turnaround in fixing security bugs
--------------------
of its rapid turnaround in fixing security bugs

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+
+     Continuing with the healthy approach of avoiding the introduction of
+     unnecessary dependency for the safety claims, the frugal designer will
+     recognise that the last thing one wants to do is to rely on Linux to
+     configure the watchdog, not only because of all the unnecessary safety
+     requirements placed on Linux, but also because such an approach would
+     leave a risk of having the system in an undefined state, potentially
+     indefinitely, should the linux boot fail before configuring successfully
+     the watchdog.
+
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:235]: The above sentence could be misleading IMO.
In fact it could be perfectly fine to have an initial config of the WTD done in firmware and looking after the boot phase, while Linux can be used to re-configure the wtd in a second phase...

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+    an existing non-safe system design into a safety qualifiable one.
+
+  * **Other Exception Context**
+ 
+    As discussed in the "Checklist" document, on a modern SoC it is not
+    uncommon to find out that the kernel is not the most privileged
+    execution context, and that other contexts are capable of interfering
+    with it in undetectable (by the kernel) ways. And the only solution
+    is likely to be requiring that these contexts must be safety qualified.
+    The hypervisor is a good example of such scenario.
```
Comment (paolonig) [inline Contributions/Using_Linux_in_a_Safe_System.md:315]: Maybe for completeness I would also mention the Trusted Execution Environment...

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+* **Application Monitoring**
+
+  Monitoring is a simple approach to detection of interference that affects
+  component with safety requirement.
+  
+  It can save from extensive hardening.
+
+  However two conditions must be met:
+  1. It must be possible to express a simple check, in a way that can repeatedly,
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:140]: express a simple check is not clear to me.
The check can be complex, I would say must be rigorous. Simple or complex are subjectives.
Are you referring to the output of the check that should be simple?

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+     reliably and unequivocally produce a YES/NO verdict about the health parameter
+     of interest
+  3. The observation whether the condition is met (or not) can be performed from a
+     context having same or better safety qualification.
+     And the context is able to take whatever action might be dictated by the
+     specific policy in place.
+
+  Assuming that the these two conditions are met (e.g. the system contains either a
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:148]: Assuming that the these two conditions are met
 should be
  Assuming that these two conditions are met

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+  watchdog or an equivalent component -- more on this later) then the opportunistic
+  approach is to identify an high-level activity that can be tied to the monitor.
+
+  A very crude example of this validation could be to require that application
+  responsible of performing safety relevant operations would ping a watchdog,
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:153]: Is the application that ping the watchdog?

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+  periodically.
+
+  As long as the application pings the watchdog within the set deadline, the watchdog
+  does not trigger any exceptional operation.
+
+  The example is crude, but it illustrates how - for example - all the internal linux
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:159]: should we refer to linux as Linux evrywhere?

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+     *  the initialisation of the watchdog can be considered sufficiently safe
+        (same or better safety level than the required one)
+     *  it must be impossible for a single failure in the system to make it
+        ineffective; this case covers:
+        *  somehow a failure either disables or relaxes either the time
+           constraint of watchdog or action-upon-expiring
+        *  somehow a failure triggers one or more pings to the watchdog,
+           when instead it should expire
+
+     Note: it is safe to allow for a more restrictive re-configuring of the
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:223]: Does that apply anytime?

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+     watchdog, because even if it happened accidentally, it would only speed
+     up the expiration of the predefined timing period.
+
+     Continuing with the healthy approach of avoiding the introduction of
+     unnecessary dependency for the safety claims, the frugal designer will
+     recognise that the last thing one wants to do is to rely on Linux to
+     configure the watchdog, not only because of all the unnecessary safety
+     requirements placed on Linux, but also because such an approach would
+     leave a risk of having the system in an undefined state, potentially
+     indefinitely, should the linux boot fail before configuring successfully
```
Comment (pellecchialuigi) [inline Contributions/Using_Linux_in_a_Safe_System.md:233]: here also linux -> Linux

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (paolonig) [reply Contributions/Using_Linux_in_a_Safe_System.md:86]: Linux is an open source project, not a product. So any SW Vendor or Integrator could perform the required qualification activities in a specific context or assumed context and your statement would not hold true anymore.
Whereas the fact that Linux was not developed according to a safety standard is a fact that would still hold true regardless of any qualification activities and regardless of any specific usage....

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 2 above)
Comment (paolonig) [reply Contributions/Using_Linux_in_a_Safe_System.md:123]: I see, ok

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (reiterative) [reply Contributions/Using_Linux_in_a_Safe_System.md:86]: I would reword this statement as: 

"Linux is not intrinsically safe"

This should not be a controversial statement, as it does not imply that Linux cannot be considered "safe enough" for use in some defined role in some defined application in some defined system context, nor do I believe that it is misleading with regard to the nature of Linux or its possible uses.

However, the phrasing does invite a different interpretation:

* "Intrinsically not safe" implies that "not safe"-ness is a natural or defining characteristic of Linux
* "Not intrinsically safe" implies that safety is not a natural or defining characteristic of Linux

I assumed that you intended to convey the second of these, Igor?

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 1 above)
Comment (reiterative) [reply Contributions/Using_Linux_in_a_Safe_System.md:113]: Gab, what about:

"However, adding additional components to a system can also increase the costs of production or maintenance, or make the associated safety arguments more difficult to follow, so it will be necessary to find the appropriate balance for any given system."

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 17 above)
Comment (reiterative) [reply Contributions/Using_Linux_in_a_Safe_System.md:140]: I think "simple" is redundant here: the important aspects are that it can be consistently and reliably performed, and that it produces an unequivocal result.

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 19 above)
Comment (reiterative) [reply Contributions/Using_Linux_in_a_Safe_System.md:153]: "...to require that an application responsible for performing safety-relevant operations must ping a watchdog periodically."

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 20 above)
Comment (reiterative) [reply Contributions/Using_Linux_in_a_Safe_System.md:159]: Yes please, let's be consistent in the capitalisation.

File: Contributions/Using_Linux_in_a_Safe_System.md (hunk 8 above)
Comment (reiterative) [reply Contributions/Using_Linux_in_a_Safe_System.md:162]: "...components involved with booting, loading and executing the application might arguably not need to be subject to any safety requirement, provided that the final product of their operation can be monitored to ensure that it satisfies the required criteria."

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+  system designs.
+
+* **Linux is not intrinsically safe**
+
+  While Linux has many remarkable properties, safety is not one of them, which means
+  that it won't be possible to just allocate safety requirements to Linux, and expect
+  them to be automatically satisfied. Quite the opposite, actually.
+
+  Having to cope with safety requirements assigned to Linux will involve a price to
+  be paid, in one form of another.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:128]:
```suggestion
  be paid, in one form or another.
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 24]
```diff
@@ -0,0 +1,486 @@
...
+
+  The problem with said cost is that it might *also* affect negatively the goals
+  that led to the selection of Linux, in the first place. For example, if Linux was
+  selected because of its a rapid turnaround in fixing security bugs, selecting
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:132]:
```suggestion
  selected because of its rapid turnaround in fixing security bugs, selecting
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 25]
```diff
@@ -0,0 +1,486 @@
...
+  an approach that would introduce an lengthy qualification process would undermine
+  the advantage of the rapid turnaround.
+
+* **Aggressively placing safety requirements on anything but (components of) the Linux kernel**
+
+  The natural consequence of the previous point is that - given a specific set of
+  safety goals for the system as a whole - one should explore if components other
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:139]:
```suggestion
  safety goals for the system as a whole - one should explore whether components other
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 26]
```diff
@@ -0,0 +1,486 @@
...
+  *  much greater probability for the component to be simpler (and thus easier to qualify)
+  *  opportunity to have a better control over the component (e.g. use a software
+     architecture that is better suited for safety, formalised processes, etc.)
+  *  decouple release cycles of the Linux kernel from the safety component
+  *  other non-technical reasons (e.g. different licensing)
+
+  A specific application or use case, will have a broader set of requirements,
+  like Total Cost of Ownership, Mean Time Before Failure, weight, repairability,
+  just to mention a few, and this can and will affect the decision of what's the
+  optimal compromise between allocating specific safety requirements on either
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:154]:
```suggestion
  just to mention a few, and this can and will affect the decision of what constitutes
  an optimal compromise between allocating specific safety requirements on either
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 27]
```diff
@@ -0,0 +1,486 @@
...
+  Linux or other components.
+
+  Placing safety requirements on a software component means that also the hardware
+  executing said software will be involved in supporting the associated safery claim.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:158]:
```suggestion
  executing said software will be involved in supporting the associated safety claim.
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 28]
```diff
@@ -0,0 +1,486 @@
...
+  Safety qualified hardware tends - unsurprisingly - to be more expensive, so
+  it might be more cost effective to place safety requirements on a smaller, cheaper
+  component, than on processor executing Linux.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:161]:
```suggestion
  component, than on the processor executing Linux.
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 29]
```diff
@@ -0,0 +1,486 @@
...
+
+  It really depends on the application. 
+
+## **Examples of Safety-oriented System Design Concepts**
+
+**Premise:** applying the initial consideration from the previous sections, it is
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:167]:
```suggestion
**Premise:** Applying the initial consideration from the previous sections, it is
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 30]
```diff
@@ -0,0 +1,486 @@
...
+necessary to define what is required at system level in terms of both availability
+and integrity.
+
+The simpler - and most common - case is to require detection of interference
+affecting components that are recipient of safety requirements.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:172]:
```suggestion
affecting components that are the recipients of safety requirements.
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 31]
```diff
@@ -0,0 +1,486 @@
...
+pre-established policy, to ensure adequate availability
+(e.g. powering off, restarting, onlining of backup system, etc.)
+
+As mentioned earlier, guaranteeing availability can be much more difficult
+and/or costly.
+
+* **Application Monitoring**
+
+  Monitoring is a simple approach to detection of interference that affects
+  components with safety requirement.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:186]:
```suggestion
  components with safety requirements.
```

File: Contributions/Using_Linux_in_a_Safe_System.md [hunk 32]
```diff
@@ -0,0 +1,486 @@
...
+  
+  It can save from extensive hardening.
```
Comment (reiterative) [inline Contributions/Using_Linux_in_a_Safe_System.md:188]:
```suggestion
  It can avoid the need for more extensive hardening.
```

**Review by reiterative — APPROVED** [review]
Approved, as all issues raised have been addressed, and all suggestions incorporated.
